esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,558 @@ int CEspHttpServer::onGet()
+{
+    m_response->setContentType("text/html; charset=UTF-8");
+    m_response->setStatus(HTTP_STATUS_OK);
+    m_response->send();
+}
+
+unsigned CEspHttpServer::createHTTPSession(EspAuthRequest& authReq, IPropertyTree* domainSessions, const char* _loginURL)
+{
+    CDateTime now;
+    now.setNow();
+    time_t createTime = now.getSimple();
+
+    StringBuffer peer;
+    VStringBuffer idStr("%s_%ld", m_request->getPeer(peer).str(), createTime);
+    unsigned sessionID = hashc((unsigned char *)idStr.str(), idStr.length(), 0);
+    ESPLOG(LogMax, "New sessionID <%d> at <%ld> in createHTTPSession()", sessionID, createTime);
+
+    Owned<IPropertyTree> ptree = createPTree();
+    ptree->addProp(PropSessionNetworkAddress, peer.str());
+    ptree->addPropInt64(PropSessionID, sessionID);
+    ptree->addPropInt(PropSessionState, HTTPSS_new);
+    ptree->setPropInt64(PropSessionCreateTime, createTime);
+    ptree->setPropInt64(PropSessionLastAccessed, createTime);
+
+    if (!isEmptyString(_loginURL))
+        ptree->addProp(PropSessionLoginURL, _loginURL);
+    else
+    {
+        StringBuffer loginURL = authReq.httpPath;
+        if (authReq.requestParams && authReq.requestParams->hasProp("__querystring"))
+            loginURL.append("?").append(authReq.requestParams->queryProp("__querystring"));
+        if (!loginURL.isEmpty() && streq(loginURL.str(), "/WsSMC/"))
+            ptree->addProp(PropSessionLoginURL, "/");
+        else
+            ptree->addProp(PropSessionLoginURL, loginURL.str());
+    }
+
+    if (domainSessions)
+    {
+        domainSessions->addPropTree(PathSessionSession, LINK(ptree));
+        return sessionID;
+    }
+
+    Owned<IRemoteConnection> conn = querySDS().connect(authReq.authBinding->getDomainSessionSDSPath(), myProcessSession(), RTM_LOCK_WRITE, SESSION_SDS_LOCK_TIMEOUT);
+    if (!conn)
+        throw MakeStringException(-1, "Failed to connect SDS DomainSession.");
+
+    IPropertyTree* root = conn->queryRoot();
+    if (!root)
+        throw MakeStringException(-1, "Failed to get SDS DomainSession.");
+
+    root->addPropTree(PathSessionSession, LINK(ptree));
+    return sessionID;
+}
+
+IPropertyTree* CEspHttpServer::readAndCleanDomainSessions(EspHttpBinding* authBinding, void* _conn, time_t timeNow)
+{
+    if (!_conn)
+        throw MakeStringException(-1, "Invalid SDS connection.");
+
+    IRemoteConnection* conn = (IRemoteConnection*) _conn;
+    IPropertyTree* root = conn->queryRoot();
+    if (!root)
+        throw MakeStringException(-1, "Failed to get SDS DomainSession.");
+
+    if (authBinding->getSessionTimeoutSeconds() < 0)
+        return root;
+
+    //Removing HTTPSessions if timed out
+    Owned<IPropertyTreeIterator> iter = root->getElements(PathSessionSession);
+    ForEach(*iter)
+    {
+        IPropertyTree& item = iter->query();
+        if (timeNow - item.getPropInt64(PropSessionLastAccessed, 0) >= authBinding->getSessionTimeoutSeconds())
+            root->removeTree(&item);
```

Comment:
I am not sure, but this may cause problems.. deleting elements that you are iterating over.
Normal practice in that case is to add to an array, then delete in 2nd pass.